### PR TITLE
Preliminary changes for kmeans-based anomaly detection.

### PR DESCRIPTION
**Pull request by jcjimenez**
For review purposes at the moment - DO NOT MERGE.

For the purposes of this explanation, consider the data as having 2 features, topic count and time period. However, in practice, two more features are taken into account: the pipeline key hash value (e.g. the hash value for `"twitter"`) and the external source id hash value (e.g. `"HamillHimself"`)

This implementation relies on first transforming raw time series data like the following:

topic_count_to_date

Into a spectrum/histogram whose data is separated into Fortis' existing period types (`day`, `week`, `month`, `year`). The following day-of-week spectrum (x axis 1: Monday, 2: Tuesday etc.) is an example of the data passed into the k-means cluster model for fitting:

topic_count_by_day_of_week

Such a transformation is then passed through a kmeans cluster whose `k` value is proportional to the number of values in the x axis. Also, the feature values in the x axis are multiplied by a factor large enough to help spread out the clusters.

Lastly, when a new topic count (`ComputedTile`) is encountered, it is passed through each of the k-means models, and if its cost is greater than the model cost (sum of squared distances of points to their nearest center), it is considered an anomaly for that period.

Also, I just watched this video and noticed they are doing something similar to what I ended up doing (histogram-based approach with DBSCAN):

https://www.youtube.com/watch?v=wacqvGVXuxc

### CODE REVIEW

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala]: Do not be spooked by these hard-coded, period-specific, values. I'm in the process of generalizing this for the different period types. Some period types, like `day`, will have multiple models (e.g. daily for day of the week, day of the month, and day of the year).

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TrustedSource.scala
Comment (c-w) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TrustedSource.scala]: Nit: elsewhere we use this style for case classes:

```scala
case class TrustedSource(
  pipelinekey: String,
  externalsourceid: String,
  sourcetype: String,
  rank: Int,
  insertion_time: Long
) extends Serializable
```

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TopicCount.scala
Comment (c-w) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TopicCount.scala]: Nit: space after colon.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/MLModel.scala
Comment (c-w) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/MLModel.scala]: Apache-Commons has a nice [helper](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/SerializationUtils.html#serialize-java.io.Serializable-) for this serialization and deserialization:

```scala
import org.apache.commons.lang.SerializationUtils.serialize;
return serialize(modelObj);
```

In case you want to use this:

```scala
// https://mvnrepository.com/artifact/commons-lang/commons-lang
libraryDependencies += "commons-lang" % "commons-lang" % "2.6"
```

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala
Comment (c-w) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala]: Do we need this call given that KMeans is unsupervised?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala
Comment (c-w) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelInputDStream.scala]: The old Java Calendar API is [pretty broken](https://stackoverflow.com/q/1571265/3817588). I'd recommend against using it and use the [Java8 Joda-time API](https://docs.oracle.com/javase/8/docs/api/java/time/ZonedDateTime.html#getDayOfMonth--) instead which is safer and less verbose:

```java
ZonedDateTime.ofInstant(Instant.ofEpochMilli(topicCount.periodstartdate), ZoneId.of("UTC")).getDayOfMonth()
```

**Review by c-w — APPROVED** [review]
Commenting only on code, not on the overall model/approach (will sync up in person).

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala
Comment (Smarker) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala]: do you need `.toDouble` here?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelParameter.scala
Comment (Smarker) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansModelParameter.scala]: Maybe you could make the k for kmeans a constant to make it a little more clear?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala
Comment (Smarker) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala]: I am a little confused by this if statement's condition: `computedCost < computedCost`. I believe this would always be false.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala
Comment (Smarker) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala]: Nit: spacing for =>

**Review by Smarker — CHANGES_REQUESTED** [review]
I think it LGTM, but there's one if statement I feel seems off and might need to be changed before merging.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala
Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala]: I don't think it's necessary - at least, in my testing, I saw Scala automatically doing the right thing, e.g.

```
scala> val a: Int = 30
a: Int = 30

scala> val b: Double = a
b: Double = 30.0
```

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala
Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala]: Woah, good catch! I flipped the condition and ended up using the same variable on both sides. It should read like:

```
        if (computedCost < modelCost) Seq()
```

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala
Comment (Smarker) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedAnomalyKmeansInputRecord.scala]: ok cool

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala
Comment (Smarker) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/analyzer/timeseries/ComputedTileToKmeansComputedAnomaly.scala]: oh haha I see. :)

**Review by Smarker — APPROVED** [review]

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TopicCount.scala
Comment (noodlefrenzy) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TopicCount.scala:4]: just curious on the lack of camelCasing - is that standard on the project?

Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/TopicCount.scala:4]: Yup, I've been following the pattern for the other Fortis entities.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala
Comment (noodlefrenzy) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala:28]: Good commenting, but might consider the ability to parameterize number of buckets to allow easier experimentation.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala
Comment (noodlefrenzy) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala:31]: I'm a big non-fan of magic constants, so I'd either consider factoring this into a well-named function or some sort of alternate approach.

**Review by noodlefrenzy — APPROVED** [review]
Approach seems reasonable for what it's doing, although I'm not totally convinced that anomalies detected in this fashion are true anomalies because I'm not convinced that the data is seasonal in this way. I'd recommend experimenting and validating that detected anomalies make sense before wide deployment.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala
Comment (jcjimenez) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/statistics/ComputedAnomalyKmeansModelParameter.scala:31]: Good point.

**Comment by jcjimenez** [comment]
Closing and doing some more research on data seasonality (or lack of).

**Pull request closed without merging**
